src/main/java/com/google/sps/servlets/AuthServlet.java

```diff
@@ -82,24 +88,73 @@ private static String convertToJson(UserAuth userAuth) {
   
   /**
    * If the user is not in database already, add them.
+   * This method returns the Entity object in the database (or the newly-created
+   * user Entity, if none previously existed).
    */
-  public void addUserToDatabase(String email) {
+  public static Entity addUserToDatabase(String email) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Only add User to database if they are not already present there.
+    Entity userEntity = getUserEntityFromEmail(email);
+    if (userEntity == null) {
+      // User is not in database, create and add user object.
+      User newUser = new User(email);
+      Entity newUserEntity = newUser.buildEntity();
+      datastore.put(newUserEntity);
+
+      // Return the new user object.
+      return newUserEntity;
+    }
+
+    // Return the user already in the database.
+    return userEntity;
+  }
+
+  /**
+   * Get the Entity object of the current user that is signed in. If not already
+   * in database, this method adds them and returns the newly-added Entity object.
+   * If no user is signed in, return null.
+   */
+  public static Entity getCurrentUserEntity() {
+    UserService userService = UserServiceFactory.getUserService();
+
+    // Return null if no user is logged in.
+    if (!userService.isUserLoggedIn()) {
+      return null;
+    }
+
+    // Get user email from UserService.
+    String email = userService.getCurrentUser().getEmail();
+
+    // Get user from datastore; if not present, add the user and return that.
+    Entity userEntity = getUserEntityFromEmail(email);
+    if (userEntity == null) {
+      Entity newUserEntity = addUserToDatabase(email);
+      return newUserEntity;
+    }
+    return userEntity;
+  }
+
+  /**
+   * Return Entity object of user from database using email, or null if user 
+   * email is not in db.
+   */
+  public static Entity getUserEntityFromEmail(String email) {
```

Comment:
Will this method be used anywhere other than the tests? I'm wondering when a client would rather use this than the addUserToDatabase method.

Reply:
The `addUserToDatabase` method (which is now named `getOrCreateUserInDatabase`) **calls** this method, `getUserEntityFromEmail`, so this is used there.

Reply:
ah sorry, I should have asked that differently. I'm wondering if this method can be private? or if you're expecting clients of this AuthServlet class to want to call getUserEntityFromEmail directly? If not then you could mark this protected (I think that'll still allow it to be called by tests) and/or add a comment indicating that this has non-private visibility only for use in tests.

Reply:
Oh I see -- yes, I think it's only used in tests! I'll confirm this and then change it, thanks for the note.

Reply:
Actually, I tried changing the access level to **protected**, but the tests can't access it then -- it throws an error when it tries to test it, and says `"getUserEntityFromEmail(java.lang.String) has protected access in com.google.sps.servlets.AuthServlet"`. This means I'll leave this as-is.